File: src/web/src/lib/ansi.ts
// Terminal programs (tsc, npm, git, …) colour their output with ANSI escape codes.
// The ESC byte is invisible in a <pre>, so those codes would otherwise leak into the
// transcript as literal "[41m"/"[0m" garbage. Strip them for display.
//
// The pattern only matches real ESC-prefixed CSI sequences, so a user's literal
// "arr[0]" or a typed "[41m" (no ESC) is left untouched.
const ANSI_RE = /\x1b\[[0-9;?]*[ -/]*[@-~]/g;

export function stripAnsi(text: string): string {
  return text.includes('\x1b') ? text.replace(ANSI_RE, '') : text;
}

File: src/runner-go/ansi.go
package main

import (
	"regexp"
	"strings"
)

// Engines colour their stderr: codex's tracing layer writes "\x1b[31mERROR\x1b[0m
// \x1b[2mcodex_core::util\x1b[0m: …" on every line it logs. Nothing downstream is a terminal —
// the ESC byte is invisible in HTML and in a SwiftUI Text, so what survives is the literal
// "[31m"/"[2m"/"[0m" around every word, on a line a client already renders in red.
//
// Stripped here, where the stderr of every engine is read, so neither the stored run_event nor
// any of the three clients has to carry them.
//
// Matches real ESC-prefixed CSI sequences only, so a log line's own "arr[0]" survives untouched.
// Mirrors src/web/src/lib/ansi.ts, which still has to run on every event already stored.
var ansiCSI = regexp.MustCompile("\x1b\\[[0-9;?]*[ -/]*[@-~]")

func stripANSI(s string) string {
	if !strings.Contains(s, "\x1b") {
		return s // the common case: a scan, not a rewrite
	}
	return ansiCSI.ReplaceAllString(s, "")
}
